### PR TITLE
Added Bitcoin Receiver & Transaction objects; support multiple charge sources

### PR DESCRIPTION
**Pull request by maxwells**
Added Bitcoin Receiver & Transaction objects

### CODE REVIEW

**Comment by maxwells** [comment]
WIP

**Comment by maxwells** [comment]
r? @zenazn 
cc @cosn @raycmorgan 

Happy to chat context/concerns IRL!

File: bitcoin_receiver.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+// BitcoinReceiverParams is the set of parameters that can be used when creating a BitcoinReceiver.
+// For more details see https://stripe.com/docs/api/#create_bitcoin_receiver.
+type BitcoinReceiverParams struct {
+	Params
+	Amount   uint64
+	Currency Currency
+	Desc     string
+	Email    string
+	Meta     map[string]string `json:"metadata"`
```
Comment (cosn) [inline bitcoin_receiver.go]: I think this is already on Params

File: bitcoin_receiver.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	RefundAddress         string                  `json:"refund_address"`
+	BitcoinUri            string                  `json:"bitcoin_uri"`
+	Meta                  map[string]string       `json:"metadata"`
+	Email                 string                  `json:"email"`
+	Payment               string                  `json:"payment"`
+	Customer              string                  `json:"customer"`
+	Transactions          *BitcoinTransactionList `json:"transactions"`
+}
+
+func (br BitcoinReceiver) PaymentType() string {
```
Comment (cosn) [inline bitcoin_receiver.go]: Comment?

File: card.go [hunk 3]
```diff
@@ -142,6 +142,10 @@ func (c *CardParams) AppendDetails(values *url.Values, creating bool) {
 	}
 }
 
+func (c Card) PaymentType() string {
```
Comment (cosn) [inline card.go]: Comment?

File: payment_source.go [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package stripe
```
Comment (cosn) [inline payment_source.go]: Can we please add comments for this whole file?

File: payment_source.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+	Card  *CardParams
+}
+
+type PaymentSource struct {
+	ID     string `json:"id"`
+	Object string `json:"object"`
+}
+
+type Chargeable interface {
+	PaymentType() string
```
Comment (cosn) [inline payment_source.go]: This isn't being invoked for anywhere ... who is this meant for?

File: payment_source.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package stripe
+
+type SourceParams struct {
+	Params
+	ID    string
+	Token string
+	Card  *CardParams
+}
+
+type PaymentSource struct {
```
Comment (cosn) [inline payment_source.go]: I don't see this being used anywhere

**Comment by cosn** [comment]
Also I'm assuming `make test` passed after these changes?

Version should be bumped as well (4.4.0) and we should add an entry in the changelog

**Comment by zenazn** [comment]
@cosn — okay if you're the primary reviewer on this? You know this code a lot better than me.

File: bitcoin_receiver.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+// property may be an id or the full struct if it was expanded.
+func (br *BitcoinReceiver) UnmarshalJSON(data []byte) error {
+	type bitcoinReceiver BitcoinReceiver
+	var r bitcoinReceiver
+	err := json.Unmarshal(data, &r)
+	if err == nil {
+		*br = BitcoinReceiver(r)
+	} else {
+		// the id is surrounded by "\" characters, so strip them
+		br.ID = string(data[1 : len(data)-1])
```
Comment (zenazn) [inline bitcoin_receiver.go]: This isn't right—the part between the `"`'s might contain escape characters or similar. It's better to `json.Unmarshal` into a string type and let the JSON package do all the hard work for you.

I get the feeling that this code was dutifully cargo-culted from somewhere else, though, so perhaps this deserves to be fixed in a separate PR.

**Comment by cosn** [comment]
yeah happily

File: payment_source.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	Display() string
+}
+
+// PaymentSource describes the payment source used to make a Charge.
+// The Type should indicate which object is fleshed out (eg. BitcoinReceiver or Card)
+// For more details see https://stripe.com/docs/api#retrieve_charge
+type PaymentSource struct {
+	Type            string `json:"object"`
+	ID              string `json:"id"`
+	Card            *Card
```
Comment (zenazn) [inline payment_source.go]: should these have `json:"-"`?

File: payment_source.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+func (s *PaymentSource) UnmarshalJSON(data []byte) error {
+	type source PaymentSource
+	var ss source
+	err := json.Unmarshal(data, &ss)
+	if err == nil {
+		*s = PaymentSource(ss)
+
+		switch s.Type {
+		case paymentsource.BitcoinReceiver:
+			s.BitcoinReceiver = &BitcoinReceiver{}
```
Comment (zenazn) [inline payment_source.go]: I'm pretty sure `encoding/json` will initialize these for you—no need to do it by hand.

File: paymentsource/types.go [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package paymentsource
+
+const (
+	BitcoinReceiver = "bitcoin_receiver"
```
Comment (zenazn) [inline paymentsource/types.go]: should these be typed constants? (do they need to be in a separate package?)

File: payment_source.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+// Interface to provide a human readable representation of a struct
+type Displayer interface {
+	Display() string
+}
+
+// PaymentSource describes the payment source used to make a Charge.
+// The Type should indicate which object is fleshed out (eg. BitcoinReceiver or Card)
+// For more details see https://stripe.com/docs/api#retrieve_charge
+type PaymentSource struct {
```
Comment (zenazn) [inline payment_source.go]: I take it you ended up doing this over the `interface` strategy?

Comment (maxwells) [inline payment_source.go]: Yeah -- I didn't really love that in order to access the common properties ("id" & "object"), end users would need a type assertion or we would provide a bunch of redundant getters to satisfy the interface. I also got to move the unmarshaling of the Card/BitcoinReceiver structs into the PaymentSource itself out of Charge, which ~broke some level of encapsulation (https://github.com/stripe/stripe-go/commit/20d58c60d5bac016bb3a13d5775443d3f1c2ad06#diff-04f5f707872b10c0ab2a591bceddbf5eL79). I may well have missed a better approach to unmarshaling into an interface property, though.

File: payment_source.go (hunk 8 above)
Comment (maxwells) [inline payment_source.go]: It's definitely redundant data, but allows the bindings user to get the common information (ID, Type) without having to inspect the child property of that type. (in `func (s *PaymentSource) UnmarshalJSON(data []byte) error` it unmarshals to the PaymentSource itself and then the appropriate child property)

File: paymentsource/types.go (hunk 10 above)
Comment (maxwells) [inline paymentsource/types.go]: yes, they should be typed!

My thinking for putting these into the paymentsource package was that the terms I wanted were already defined in the top level stripe package. I could certainly do `BitcoinReceiverType` or `CardSourceType` or something to that effect and define them in the top level `payment_source.go` file -- is that a more reasonable approach?

Thanks much for looking this over, btw.

File: payment_source.go (hunk 9 above)
Comment (maxwells) [inline payment_source.go]: you're correct - thanks for the tip

File: payment_source.go (hunk 8 above)
Comment (zenazn) [inline payment_source.go]: err, as in the following JSON blob will deserialize in an unexpected way:

```
{
    "type": "card",
    "card": {
        "number": "12345"
    },
    "expiry": "1/23"
}
```

(a similar thing will be true for the case of JSON marshaling)

**Comment by maxwells** [comment]
@cosn Yeah all tests pass provided I use test keys that have a number of charges that fit into a `uint16`.

```go
type ListMeta struct {
	Count uint16 `json:"total_count"` // <-- Unmarshaling fails otherwise
	More  bool   `json:"has_more"`
	URL   string `json:"url"`
}
```

File: payment_source.go (hunk 8 above)
Comment (maxwells) [inline payment_source.go]: Gotcha. Yes. Forgot how github line comments work and thought you were referring to L28-29. I'll get that patched up

**Comment by cosn** [comment]
Cool, let me know when all the udpates are done and I'll go through it again

File: bitcoin_receiver.go [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+	"fmt"
+)
+
+// BitcoinReceiverListParams is the set of parameters that can be used when listing BitcoinReceivers.
+// For more details see https://stripe.com/docs/api/#list_bitcoin_receivers.
+type BitcoinReceiverListParams struct {
+	ListParams
+	NotFilled  bool
+	NotActive  bool
+	Uncaptured bool
```
Comment (cosn) [inline bitcoin_receiver.go]: `NotFilled, NotActive, Uncaptured bool`

File: bitcoin_receiver.go [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// BitcoinReceiverParams is the set of parameters that can be used when creating a BitcoinReceiver.
+// For more details see https://stripe.com/docs/api/#create_bitcoin_receiver.
+type BitcoinReceiverParams struct {
+	Params
+	Amount   uint64
+	Currency Currency
+	Desc     string
+	Email    string
```
Comment (cosn) [inline bitcoin_receiver.go]: `Desc, Email string`

File: bitcoin_transaction.go [hunk 14]
```diff
@@ -0,0 +1,47 @@
+package stripe
+
+import "encoding/json"
+
+// BitcoinTransactionListParams is the set of parameters that can be used when listing BitcoinTransactions.
+type BitcoinTransactionListParams struct {
+	ListParams
+	Receiver string
+	Customer string
```
Comment (cosn) [inline bitcoin_transaction.go]: `Receiver, Customer string`

File: bitcointransaction/client.go [hunk 15]
```diff
@@ -0,0 +1,71 @@
+// Package bitcointransaction provides the /bitcoin/transactions APIs
+package bitcointransaction
+
+import (
+	"net/url"
+
```
Comment (cosn) [inline bitcointransaction/client.go]: why newline here?

File: card.go [hunk 16]
```diff
@@ -142,6 +143,11 @@ func (c *CardParams) AppendDetails(values *url.Values, creating bool) {
 	}
 }
 
+// Human readable/displayable way of inspecting a Card
+func (c *Card) Display() string {
+	return fmt.Sprintf("%s (Last Four: %s)", c.Brand, c.LastFour)
```
Comment (cosn) [inline card.go]: `%s ending in %`

File: card.go [hunk 17]
```diff
@@ -142,6 +143,11 @@ func (c *CardParams) AppendDetails(values *url.Values, creating bool) {
 	}
 }
 
+// Human readable/displayable way of inspecting a Card
```
Comment (cosn) [inline card.go]: This should start with "Display". Also period at the end

File: payment_source.go [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+// can be described by SourceParams.
+// For more details see https://stripe.com/docs/api#create_charge
+type SourceParams struct {
+	Params
+	ID    string
+	Token string
+	Card  *CardParams
+}
+
+// Interface to provide a human readable representation of a struct
```
Comment (cosn) [inline payment_source.go]: Should begin with "Displayer"

File: paymentsource/types.go [hunk 19]
```diff
@@ -0,0 +1,6 @@
+package paymentsource
```
Comment (cosn) [inline paymentsource/types.go]: rather than putting this in its own package, can we put it in `payment_source.go` and name it `PaymentSource{BitcoinReceiver,Card}`

File: payment_source.go [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+type Displayer interface {
+	Display() string
+}
+
+// PaymentSource describes the payment source used to make a Charge.
+// The Type should indicate which object is fleshed out (eg. BitcoinReceiver or Card)
+// For more details see https://stripe.com/docs/api#retrieve_charge
+type PaymentSource struct {
+	Type            string           `json:"object"`
```
Comment (cosn) [inline payment_source.go]: Rather than `string`, can we make this a `PaymentSourceType` that just has the two consts below?

File: bitcointransaction/client.go [hunk 21]
```diff
@@ -0,0 +1,71 @@
+// Package bitcointransaction provides the /bitcoin/transactions APIs
```
Comment (cosn) [inline bitcointransaction/client.go]: period at the end

File: bitcoinreceiver/client.go [hunk 22]
```diff
@@ -0,0 +1,120 @@
+// Package bitcoinreceiver provides the /bitcoin/receivers APIs
```
Comment (cosn) [inline bitcoinreceiver/client.go]: period at the end

File: bitcoin_receiver.go [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+	RefundAddress         string                  `json:"refund_address"`
+	BitcoinUri            string                  `json:"bitcoin_uri"`
+	Meta                  map[string]string       `json:"metadata"`
+	Email                 string                  `json:"email"`
+	Payment               string                  `json:"payment"`
+	Customer              string                  `json:"customer"`
+	Transactions          *BitcoinTransactionList `json:"transactions"`
+}
+
+// Human readable/displayable way of inspecting a BitcoinReceiver
```
Comment (cosn) [inline bitcoin_receiver.go]: This should begin with "Display". Also period at the end

**Comment by cosn** [comment]
Generally looks good, just some minor nits. Once we get those cleared I'm happy to merge this (or feel free to merge yourself)
